WeScan/Common/CaptureSession.swift

```diff
@@ -29,3 +30,90 @@ final class CaptureSession {
+    enum FlashState {
+        case on
+        case off
+        case unavailable
+        case unknown
+    }
+    
+    /// Toggles the current device's flashlight on or off.
+    func toggleFlash() -> FlashState {
+        guard let device = AVCaptureDevice.default(for: AVMediaType.video), device.isTorchAvailable else { return .unavailable }
+        
+        do {
+            try device.lockForConfiguration()
+        } catch {
+            return .unknown
+        }
+        
+        defer {
+            device.unlockForConfiguration()
+        }
+        
+        if device.torchMode == .on {
+            device.torchMode = .off
+            return .off
+        } else if device.torchMode == .off {
+            device.torchMode = .on
+            return .on
+        }
+        
+        return .unknown
+    }
+}
+
+/// Extension to CaptureSession that controls auto focus
+extension CaptureSession {
+    /// Sets the camera's exposure and focus point to the given point
+    func setFocusPointToTapPoint(_ tapPoint: CGPoint) throws {
+        guard let device = AVCaptureDevice.default(for: AVMediaType.video) else { return }
+        try device.lockForConfiguration()
+        
+        if device.isFocusPointOfInterestSupported && device.isFocusModeSupported(.autoFocus) {
+            device.focusPointOfInterest = tapPoint
+            device.focusMode = .autoFocus
+        }
+        
+        if device.isExposurePointOfInterestSupported, device.isExposureModeSupported(.continuousAutoExposure) {
+            device.exposurePointOfInterest = tapPoint
+            device.exposureMode = .continuousAutoExposure
+        }
+        
+        device.unlockForConfiguration()
+    }
+    
+    /// Resets the camera's exposure and focus point to automatic
+    func resetFocusToAuto() throws {
+        guard let device = AVCaptureDevice.default(for: AVMediaType.video) else { return }
+        try device.lockForConfiguration()
+        
+        if device.isFocusPointOfInterestSupported && device.isFocusModeSupported(.continuousAutoFocus) {
+            device.focusMode = .continuousAutoFocus
+        }
+        
+        if device.isExposurePointOfInterestSupported, device.isExposureModeSupported(.continuousAutoExposure) {
+            device.exposureMode = .continuousAutoExposure
+        }
+        
+        device.unlockForConfiguration()
+    }
+    
+    /// Removes an existing focus rectangle if one exists, optionally animating the exit
+    func removeFocusRectangleIfNeeded(_ focusRectangle: FocusRectangle?, animated: Bool) {
+        guard let focusRectangle = focusRectangle else { return }
+        if animated {
+            UIView.animate(withDuration: 0.3, delay: 1.0, animations: {
+                focusRectangle.alpha = 0.0
+            }, completion: { (_) in
+                focusRectangle.removeFromSuperview()
+            })
+        } else {
+            focusRectangle.removeFromSuperview()
+        }
+    }
+}
```

Comment:
This part should be testable and was what I was aiming for. Should be doable with protocols. Let me know if you have any questions!

Reply:
Hey, not really sure how you intend this to be tested? Do you just want a ScannerVC to be created, then these methods called and stuff like the focus rectangle existing/not existing being tested?

Reply:
I think the easiest way to test this is just to create a CaptureSession instance and test that when this is called the focusRectangle is not longer in the view hierarchy.

Unfortunately the animation will be hard to test unless you disable animations gloaballly using CATranasaction so I think we should just agree to test the non-animated variant for now.

Reply:
Hopefully this can help you see ways you can make the function to be easier to test once you've written the first version of those tests :)